modules/product/src/VariationFieldInjection.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+namespace Drupal\commerce_product;
+
+use Drupal\commerce_product\Entity\ProductVariationInterface;
+use Drupal\Core\Ajax\AjaxResponse;
+use Drupal\Core\Ajax\ReplaceCommand;
+use Drupal\Core\Entity\EntityFieldManagerInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Field\BaseFieldDefinition;
+
+/**
+ * Default implementation of VariationFieldInjectionInterface.
+ */
+class VariationFieldInjection implements VariationFieldInjectionInterface {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The entity field manager.
+   *
+   * @var \Drupal\Core\Entity\EntityFieldManagerInterface
+   */
+  protected $entityFieldManager;
+
+  /**
+   * The module handler.
+   *
+   * @var \Drupal\Core\Extension\ModuleHandlerInterface
+   */
+  protected $moduleHandler;
+
+  /**
+   * Local cache for injectable field definitions.
+   *
+   * @var \Drupal\Core\Field\FieldDefinitionInterface[]
+   */
+  protected $fieldDefinitions = [];
+
+
+  /**
+   * Constructs a new VariationFieldInjector object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entity_field_manager
+   *   The entity field manager.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager, EntityFieldManagerInterface $entity_field_manager, ModuleHandlerInterface $module_handler) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->entityFieldManager = $entity_field_manager;
+    $this->moduleHandler = $module_handler;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFieldDefinitions($variation_type_id) {
+    if (!isset($this->fieldDefinitions[$variation_type_id])) {
+      $definitions = $this->entityFieldManager->getFieldDefinitions('commerce_product_variation', $variation_type_id);
+      $definitions = array_filter($definitions, function ($definition) {
+        /** @var \Drupal\Core\Field\FieldDefinitionInterface $definition */
+        $name = $definition->getName();
+        if ($definition instanceof BaseFieldDefinition && !in_array($name, $this->allowedBaseFields())) {
+          return FALSE;
+        }
+        return (strpos($name, 'attribute_') === FALSE);
+      });
+
+      $this->fieldDefinitions[$variation_type_id] = $definitions;
+    }
+
+    return $this->fieldDefinitions[$variation_type_id];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRenderableFields(ProductVariationInterface $variation, $view_mode = 'default') {
```

Comment:
I like the getter more, as it is not actually rendering the fileds, but collecting the renderable fields